core/src/services/s3/lister.rs

```diff
@@ -125,12 +126,25 @@ impl oio::PageList for S3Lister {
             }
             meta.set_content_length(object.size);
 
-            // object.last_modified provides more precious time that contains
+            // object.last_modified provides more precise time that contains
             // nanosecond, let's trim them.
             meta.set_last_modified(parse_datetime_from_rfc3339(object.last_modified.as_str())?);
 
+            if path == self.path {
+                ctx.path_added = true;
+            }
             let de = oio::Entry::with(path, meta);
             ctx.entries.push_back(de);
+
+            prefix_existing = true;
+        }
+
+        // if the path is dir, but hasn't created yet and exists only as the prefix of a key,
```

Comment:
Hi, I think it's fine for s3 to not return dir itself if it doesn't exist. Our list behavior is `returning the dir itself if it does exist` which make it super useful while users perform list and delete.

Reply:
got it. This way, the behavior will be consistent with S3 itself.
however, we need to create every dir in the tests, or the tests will pass for FS, but fail for S3.

Reply:
I think we can test in this way. For example, give `x/x/x/x`.

We can manually call `create_dir("x/x/")` and check `x/x/` to ensure it has been created. However, for `x/x/x/`, we won't check it; we'll just consider it an implementation detail.

What do you think?

Reply:
yes, then we can do `list("x/x/")` or `list("x/x/")`; but can't do `list("x/)` in the tests. which dir should be created depends on the test case, we don't have to create every dir. I've checked all test cases, and the modifications are minimal.